Codegen support common subexpression replacement

peabody124 opened the pull request:
This makes an extension to the Routine class in codegen to allow adding Result
objects to the local_vars field. These are then output as C assignments in the
CCodeGen class in the _declare_locals method. This allows defining routines with
local variable calculations and writing them correctly.

Next, it extends the CCodeGen object so that if the cse field is True, it will perform
CSE on the expressions passed in. It keeps the correct set of input and output
arguments while doing this. It correctly handles the case of MatrixSymbol input and
output arguments.

Provides some work towards #11038

peabody124 committed:
Codegen: expose OutputArgument

peabody124 committed:
Codegen: sensible output for Result

peabody124 committed:
Codegen: routine supports local results

peabody124 committed:
Codegen: check in some work to make it use CSE

peabody124 committed:
Codegen: clean up CSE layer

peabody124 committed:
Codegen: routine method now can perform CSE

This allows making the automagically generated C code
more efficient.

peabody124 commented:
Thanks for your work on Sympy guys. This is a great package. Critical review welcomed, as I'm just learning my way around the code base. I definitely may not have covered all the usage cases of codegen and the way inputs can be represented. I was mostly coding to my own usages. If there is somewhere I should add unit tests, please point me that way and I can do so.

 Here are some simple examples to show it does what I'd expect.

A simple expression vector assignment without CSE:
![screen shot 2017-08-23 at 6 43 16 pm](https://user-images.githubusercontent.com/618346/29643164-f95ae84c-8832-11e7-930c-bffb2f5c8b29.png)

And the same equation with CSE:
![screen shot 2017-08-23 at 6 43 23 pm](https://user-images.githubusercontent.com/618346/29643174-0d8a9c72-8833-11e7-8317-ab03e0361264.png)

Note however, that if you are passing vectors as inputs and outputs you get a number of silly redefinitions. 
![screen shot 2017-08-23 at 6 45 16 pm](https://user-images.githubusercontent.com/618346/29643205-3a71dcfa-8833-11e7-92e9-93fbd4cef563.png)

This later is due to a know bug in CSE w.r.t. MatrixSymbols and still produces valid code. #11991

peabody124 committed:
Codegen: preserve dimensions of inputs with CSE

peabody124 committed:
Autowrap: use routine method for code_gen if passed

peabody124 committed:
Codegen: using CSE and lists of Equations now works

For example now this snippet generates the correct code

    a = sm.MatrixSymbol('a', 3, 1)
    out1 = sm.MatrixSymbol('out1', 3, 1)
    out2 = sm.MatrixSymbol('out2', 3, 1)
    eq1 = sm.Eq(out1, sm.Matrix([a[0],a[0]*a[1],a[0]*a[1]*a[2]]))
    eq2 = sm.Eq(out2, sm.Matrix([a[1]*a[2],a[1]*a[1],a[2]*a[2]*a[2]*a[1]*a[0]]))
    eq = [eq1, eq2]
    r = generator.routine('a', eq, argument_sequence=(a, out1, out2))

peabody124 committed:
Codegen: local variables constant when not arrays

bjodah commented:
@peabody124 thank you for this PR, it looks interesting.

It does need tests, something along these lines:
https://gist.github.com/anonymous/0f45ce576b2cd6ef4ea28c9aea445def#file-sandbox-ipynb

I also noticed that running that notebook twice (without restarting it) leads to failure (I'm guessing some class mutable attribute, e.g. a list, is being appended to). Not sure if that's new since this PR or if it's always been like that.

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -117,45 +117,37 @@
 
 class Routine(object):
     """Generic description of evaluation routine for set of expressions.
-
```
You should leave this new line be. (first line of a docstring should be a one sentence summary)

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -117,45 +117,37 @@
 
     def __init__(self, name, arguments, results, local_vars, global_vars):
         """Initialize a Routine instance.
-
```
same here

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -177,13 +169,20 @@ def __init__(self, name, arguments, results, local_vars, global_vars):
                 raise ValueError("Unknown Routine result: %s" % r)
             symbols.update(r.expr.free_symbols)
 
+        local_symbols = set([])
```
or just `set()`

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -666,6 +707,8 @@ def routine(self, name, expr, argument_sequence, global_vars):
         return Routine(name, arg_list, return_val, local_vars, global_vars)
 
+
+
```
The convention is one blank line between methods (and two blank lines between e.g. classes)

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -177,13 +169,20 @@ def __init__(self, name, arguments, results, local_vars, global_vars):
+        local_symbols = set([])
+        for r in local_vars:
+            if not isinstance(r, Result):
+                raise ValueError("Unknown Routine result: %s" % r)
```
This seems to be causing many test failures.

peabody124 replied:
ahh.. thank you, that led me to some use cases of the local_vars before. I was having a challenge understanding what they did (iterate through arrays) and I definitely broke some of that compatibility. I'll work on this.

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -552,21 +556,56 @@ def routine(self, name, expr, argument_sequence, global_vars):
+            if is_sequence(expr) and not isinstance(expr, (MatrixBase, MatrixExpr)):
+                if not expr:
+                    raise ValueError("No expression given")
+                assert all([e.is_Equality for e in expr]), "Lists of expressions must all be Equalities"
```
change use of `assert` to an explicit raise of some exception (and those should be tested as well). We try to contain the use of `assert` to test files only (since python can be instructed to ignore assert statements).

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -552,21 +556,56 @@ def routine(self, name, expr, argument_sequence, global_vars):
-        local_vars = {i.label for i in expressions.atoms(Idx)}
-
+        if self.cse:
+            assert {i.label for i in expressions.atoms(Idx)} == set(), "CSE and Indexed expressions do not play well together yet"
```
same here (use of assert)

bjodah commented on sympy/utilities/codegen.py:
```diff
@@ -859,8 +902,53 @@ def _declare_globals(self, routine):
+                t = result.get_datatype('c')
+                if isinstance(result.expr, (MatrixBase, MatrixExpr)):
+                    dims = result.expr.shape
+                    assert dims[1] == 1, "2D matrices not supported yet"
```
change not to use assert and maybe "only column vectors supported for now" (since matrices are 2D by definition)

bjodah reviewed (changes requested):
The CSE capabilities should be tested for Fortran as well.

Once this gets some tests (and our current tests are passing) I will give it another round of review.

It looks great as a first contribution by the way!

peabody124 committed:
Codegen: white space cleanup

peabody124 commented on sympy/utilities/codegen.py:
```diff
@@ -528,16 +531,17 @@ def _get_symbol(self, s):
+        self.cse = cse
 
-    def routine(self, name, expr, argument_sequence, global_vars):
+    def routine(self, name, expr, argument_sequence=None, global_vars=None):
```
btw I can easily be talked out of allowing the argument_sequence to be none as you will end up with something random. that might sufficient if you are developing code you will dump somewhere else, though.

peabody124 committed:
Codegen: replace asserts with exceptions

peabody124 commented:
BTW codegen has a lot of unit test failures with python2. All for importing cfunctions as shown below. Is sympy meant to pass with python2 or is support deprecated?

```
  File "/users/jcotton81/python/sympy/sympy/utilities/tests/test_codegen.py", line 539, in test_ccode_results_named_ordered
    argument_sequence=(x, C, z, y, A, B))
  File "sympy/utilities/codegen.py", line 2011, in codegen
    code_gen = get_code_generator(language, project, standard)
  File "sympy/utilities/codegen.py", line 1867, in get_code_generator
    return CodeGenClass(project)
  File "sympy/utilities/codegen.py", line 799, in __init__
    self.printer = printer or c_code_printers[self.standard.lower()]()
  File "sympy/printing/ccode.py", line 207, in __init__
    self.math_macros = settings.pop('math_macros', get_math_macros())
  File "sympy/printing/ccode.py", line 102, in get_math_macros
    from sympy.codegen.cfunctions import log2, Sqrt
ImportError: No module named cfunctions
```

peabody124 committed:
Codegen: routine handles both local vars as Result or symbol

Results mean that a local assignment will be defined in the method
so have both a rhs and lhs. These are generated by _declare_locals.
Symbols will be generated in other parts of code. In both cases,
they are covered variables.

peabody124 committed:
Codegen: _declare_locals should not do anything with symbols

Related to the previous commit, basic symbols passed in the routine
will be declared in other parts of the code (such as for loops).

peabody124 commented on sympy/utilities/codegen.py:
```diff
@@ -552,21 +568,58 @@ def routine(self, name, expr, argument_sequence, global_vars):
+                    raise ValueError("No expression given")
+                for e in expr:
+                    if not e.is_Equality:
+                        raise CodeGenError("Lists of expressions must all be Equalities. {} is not.".format(e))
```
tested:

```
In [34]: generator = CCodeGen(cse=True)

In [35]: generator.routine("test", [Eq(a,c),Eq(b,c)])
Out[35]: Routine('test', [InputArgument(c), OutputArgument(a, c, a), OutputArgument(b, c, b)], [], set(), set())

In [36]: generator.routine("test", [a,b])
---------------------------------------------------------------------------
CodeGenError                              Traceback (most recent call last)
<ipython-input-36-b8d01875f704> in <module>()
----> 1 generator.routine("test", [a,b])

~/python/sympy/sympy/utilities/codegen.py in routine(self, name, expr, argument_sequence, global_vars)
    577                 for e in expr:
    578                     if not e.is_Equality:
--> 579                         raise CodeGenError("Lists of expressions must all be Equalities. {} is not.".format(e))
    580                 lhs = [e.lhs for e in expr]
    581 

CodeGenError: Lists of expressions must all be Equalities. a is not.
```

peabody124 committed:
Codegen: OutputArgument print with params in same order as constructor

Having them switch in the string definition is very confusing when
you are creating "identical" variables.

peabody124 committed:
Codegen: cleaner exception description

peabody124 committed:
Codegen: Result print with params in same order as constructor

Note, slight confusing is the fact that the same parameters are passed
to Result and OutputArgument in different orders.

peabody124 commented on sympy/utilities/codegen.py:
```diff
@@ -859,8 +913,47 @@ def _declare_globals(self, routine):
+            if isinstance(result.expr, (MatrixBase, MatrixExpr)):
+                dims = result.expr.shape
+                if dims[1] != 1:
+                    raise CodeGenError("Only column vectors are supported in local variabels. Local result {} has dimensions {}".format(result, dims))
```
tested. note that this should never happen using CSE as that only generates a list of assignments. it would be easy to declare a variable here as the right side, but the code below generates c code that unwraps that into a linear array. it seems risky to trust the memory layout when generating a two-dimensional float var[N1][N2] but would probably fly.

```
In [1]: from sympy import *
f
In [2]: from sympy.utilities.codegen import *

In [3]: a = MatrixSymbol('a', 2, 2)

In [4]: b = MatrixSymbol('b', 2, 2)

In [5]: r = Routine('a', [OutputArgument(a, a, b)], [], [Result(Matrix([[0, 1], [2, 3]]),b)], set())

In [6]: generator = CCodeGen(cse=True)

In [7]: generator.write([r], "output")[0][1]
---------------------------------------------------------------------------
CodeGenError                              Traceback (most recent call last)
<ipython-input-7-271abd7d1525> in <module>()
----> 1 generator.write([r], "output")[0][1]

~/python/sympy/sympy/utilities/codegen.py in write(self, routines, prefix, to_files, header, empty)
    760                 filename = "%s.%s" % (prefix, dump_fn.extension)
    761                 contents = StringIO()
--> 762                 dump_fn(self, routines, contents, prefix, header, empty)
    763                 result.append((filename, contents.getvalue()))
    764             return result

~/python/sympy/sympy/utilities/codegen.py in dump_c(self, routines, f, prefix, header, empty)
   1001 
   1002     def dump_c(self, routines, f, prefix, header=True, empty=True):
-> 1003         self.dump_code(routines, f, prefix, header, empty)
   1004     dump_c.extension = code_extension
   1005     dump_c.__doc__ = CodeGen.dump_code.__doc__

~/python/sympy/sympy/utilities/codegen.py in dump_code(self, routines, f, prefix, header, empty)
    801             code_lines.extend(self._declare_arguments(routine))
    802             code_lines.extend(self._declare_globals(routine))
--> 803             code_lines.extend(self._declare_locals(routine))
    804             if empty:
    805                 code_lines.append("\n")

~/python/sympy/sympy/utilities/codegen.py in _declare_locals(self, routine)
    938                 dims = result.expr.shape
    939                 if dims[1] != 1:
--> 940                     raise CodeGenError("Only column vectors are supported in local variabels. Local result {} has dimensions {}".format(result, dims))
    941                 code_lines.append("{0} {1}[{2}];\n".format(t, str(assign_to), dims[0]))
    942                 prefix = ""

CodeGenError: Only column vectors are supported in local variabels. Local result Result(b, Matrix([
[0, 1],
[2, 3]]), b) has dimensions (2, 2)
```

bjodah replied:
I see. Yeah, we shouldn't make any assumptions about memory layout for generated C code.

peabody124 commented on sympy/utilities/codegen.py:
```diff
@@ -552,21 +567,58 @@ def routine(self, name, expr, argument_sequence, global_vars):
-        local_vars = {i.label for i in expressions.atoms(Idx)}
-
+        if self.cse and {i.label for i in expressions.atoms(Idx)} != set():
+            raise CodeGenError("CSE and Indexed expressions do not play well together yet")
```
```
In [12]: r = generator.routine('a', Eq(A[i], B[i])) #[OutputArgument(a, a, b)], [], [Result(Matrix([[0, 1], [2, 3]]),b)], set())
---------------------------------------------------------------------------
CodeGenError                              Traceback (most recent call last)
<ipython-input-12-14ac4c3345ea> in <module>()
----> 1 r = generator.routine('a', Eq(A[i], B[i])) #[OutputArgument(a, a, b)], [], [Result(Matrix([[0, 1], [2, 3]]),b)], set())

~/python/sympy/sympy/utilities/codegen.py in routine(self, name, expr, argument_sequence, global_vars)
    609 
    610         if self.cse and {i.label for i in expressions.atoms(Idx)} != set():
--> 611             raise CodeGenError("CSE and Indexed expressions do not play well together yet")
    612         else:
    613             # local variables for indexed expressions

CodeGenError: CSE and Indexed expressions do not play well together yet
```

bjodah commented:
Importing `.codgen.cfunctions` should work under Python 2. Are the test failures local or on travis? If the former you may want to check that your python2 environment isn't pulling in an older version of sympy, e.g. run `python2 -m pip install --user -e .` in sympy's repo. Also inspecting sympy.__file__ under pdb might be enlighting.

As for tests I meant that we need tests in our test files, see these files:
https://github.com/sympy/sympy/blob/master/sympy/utilities/tests/test_codegen.py
https://github.com/sympy/sympy/blob/master/sympy/utilities/tests/test_autowrap.py
https://github.com/sympy/sympy/blob/master/sympy/external/tests/test_autowrap.py
https://github.com/sympy/sympy/blob/master/sympy/external/tests/test_codegen.py

peabody124 committed:
Codegen: fix regression when adding asserts

At this point the below generation works.
    import numpy as np
    import sympy as sm
    from sympy.utilities.codegen import Routine, CCodeGen
    a = sm.MatrixSymbol('a', 3, 1)
    out1 = sm.MatrixSymbol('out1', 3, 1)
    out2 = sm.MatrixSymbol('out2', 3, 1)
    eq1 = sm.Eq(out1, sm.Matrix([a[0],a[0]*a[1],a[0]*a[1]*a[2]]))
    eq2 = sm.Eq(out2, sm.Matrix([a[1]*a[2],a[1]*a[1],a[2]*a[2]*a[2]*a[1]*a[0]]))
    eq = [eq1, eq2]
    generator = CCodeGen(cse=True)
    r = generator.routine('f', eq, argument_sequence=(a, out1, out2))

peabody124 committed:
Codegen: add unit test for using CSE

peabody124 commented:
@bjodah here is a sandbox with the same tests, using CSE https://gist.github.com/peabody124/2ea6155c6527d5cca313ba792c300313 that issue you saw was related to CythonCodeWrapper so i'm just using autowrap 

i also checked in a unit test for CSE generation

bjodah commented:
Looks good. +1 to merge as soon as tests pass (looks like our code quality test is complaining about trailing whitespace)

peabody124 committed:
Codegen: more whitespace fixes

peabody124 commented:
Awesome. I think I caught all the whitespace checks. I like that sympy.test("_quality"). The fun of working with new OSS projects -- learning improved workflows!

If this gets merged, I'll expand the test coverage in #13185 with a code generation case including matrix inputs.

peabody124 commented:
This passes tests now. Let me know if you want any more changes.

bjodah commented:
Thanks a lot!
